commit ec9c4215fef37da6668c4105f5ad3891aaa6527a
Author: Sabrina Dubroca <sd@queasysnail.net>
Date:   Tue Apr 25 15:56:50 2017 +0200

    ipv6: fix source routing
    
    Commit a149e7c7ce81 ("ipv6: sr: add support for SRH injection through
    setsockopt") introduced handling of IPV6_SRCRT_TYPE_4, but at the same
    time restricted it to only IPV6_SRCRT_TYPE_0 and
    IPV6_SRCRT_TYPE_4. Previously, ipv6_push_exthdr() and fl6_update_dst()
    would also handle other values (ie STRICT and TYPE_2).
    
    Restore previous source routing behavior, by handling IPV6_SRCRT_STRICT
    and IPV6_SRCRT_TYPE_2 the same way as IPV6_SRCRT_TYPE_0 in
    ipv6_push_exthdr() and fl6_update_dst().
    
    Fixes: a149e7c7ce81 ("ipv6: sr: add support for SRH injection through setsockopt")
    Signed-off-by: Sabrina Dubroca <sd@queasysnail.net>
    Reviewed-by: Hannes Frederic Sowa <hannes@stressinduktion.org>
    Signed-off-by: David S. Miller <davem@davemloft.net>

diff --git a/net/ipv6/exthdrs.c b/net/ipv6/exthdrs.c
index 25192a3b0cd7..d32e2110aff2 100644
--- a/net/ipv6/exthdrs.c
+++ b/net/ipv6/exthdrs.c
@@ -909,6 +909,8 @@ static void ipv6_push_rthdr(struct sk_buff *skb, u8 *proto,
 {
 	switch (opt->type) {
 	case IPV6_SRCRT_TYPE_0:
+	case IPV6_SRCRT_STRICT:
+	case IPV6_SRCRT_TYPE_2:
 		ipv6_push_rthdr0(skb, proto, opt, addr_p, saddr);
 		break;
 	case IPV6_SRCRT_TYPE_4:
@@ -1163,6 +1165,8 @@ struct in6_addr *fl6_update_dst(struct flowi6 *fl6,
 
 	switch (opt->srcrt->type) {
 	case IPV6_SRCRT_TYPE_0:
+	case IPV6_SRCRT_STRICT:
+	case IPV6_SRCRT_TYPE_2:
 		fl6->daddr = *((struct rt0_hdr *)opt->srcrt)->addr;
 		break;
 	case IPV6_SRCRT_TYPE_4: